### PR TITLE
refactor!: refactor code to support more signature envelopes

### PR DESCRIPTION
**Pull request by shizhMSFT**
This PR is a sum of PRs to refactor code to support more future signature envelopes like COSE.

All commits / PRs in this PR have been reviewed individually as follows.
- #71
- #73
- #76

### CODE REVIEW

**Comment by shizhMSFT** [comment]
Although the DCO check fails in this PR, the DCO checks are actually passed in the sub-PRs.

@gokarnm We can "set DCO to pass" before merging this PR.

**Comment by shizhMSFT** [comment]
**NOTE** This PR will be merged by rebase merging.

**Review by patrickzheng200 — APPROVED** [review]
I've reviewed all individual PRs of this PR, so LGTM.

File: signature/jws/jwt.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+	// CertificateChain returns the certificate chain.
+	//
+	// It should be called after calling Sign().
+	CertificateChain() ([]*x509.Certificate, error)
+
+	// PrivateKey returns the private key.
+	PrivateKey() crypto.PrivateKey
+}
+
+// remoteSigningMethod wraps the remote signer to be a SigningMethod
```
Comment (binbin-li) [inline signature/jws/jwt.go:28]: nit: missing period in the end.

**Review by priteshbandi — APPROVED** [review]
LGTM. Verified that all 3 commits were reviewed earlier as part #71 , #73 and #76

File: signature/jws/types.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+	// Lists the headers that implementation MUST understand and process.
+	Critical []string `json:"crit,omitempty"`
+
+	// The "best by use" time for the artifact, as defined by the signer.
+	Expiry *time.Time `json:"io.cncf.notary.expiry,omitempty"`
+
+	// Specifies the Notary v2 Signing Scheme used by the signature.
+	SigningScheme signature.SigningScheme `json:"io.cncf.notary.signingScheme"`
+
+	// The time at which the signature was generated. only valid when signing scheme is `notary.x509`
```
Comment (binbin-li) [inline signature/jws/types.go:49]: same

File: signature/jws/types.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+// jwsUnprotectedHeader contains the set of unprotected headers.
+type jwsUnprotectedHeader struct {
+	// RFC3161 time stamp token Base64-encoded.
+	TimestampSignature []byte `json:"io.cncf.notary.timestampSignature,omitempty"`
+
+	// List of X.509 Base64-DER-encoded certificates
+	// as defined at https://datatracker.ietf.org/doc/html/rfc7515#section-4.1.6.
+	CertChain [][]byte `json:"x5c"`
+
+	// SigningAgent used for signing
```
Comment (binbin-li) [inline signature/jws/types.go:68]: same

**Review by binbin-li — APPROVED** [review]
left some minor comments, lgtm

File: signature/jws/jwt.go (hunk 1 above)
Comment (shizhMSFT) [reply signature/jws/jwt.go:28]: Let's address them in other PRs.

File: signature/jws/types.go (hunk 2 above)
Comment (shizhMSFT) [reply signature/jws/types.go:49]: Let's address them in other PRs.

File: signature/jws/types.go (hunk 3 above)
Comment (shizhMSFT) [reply signature/jws/types.go:68]: Let's address them in other PRs.

**Pull request merged**